yarn-project/simulator/src/avm/errors.ts

```diff
@@ -109,13 +109,21 @@ export class AvmRevertReason extends ExecutionError {
  * @param nestedError - the error that caused this one (if this is not the root-cause itself)
  */
 function createRevertReason(message: string, context: AvmContext, nestedError?: AvmRevertReason): AvmRevertReason {
+  // TODO(https://github.com/AztecProtocol/aztec-packages/issues/8985): Properly fix this.
+  // If the function selector is the public dispatch selector, we need to extract the actual function selector from the calldata.
+  // We should remove this because the AVM (or public protocol) shouldn't be aware of the public dispatch calling convention.
+  let functionSelector = context.environment.functionSelector;
+  const internalCallStack = context.machineState.internalCallStack;
+  if (functionSelector.toField().equals(new Fr(PUBLIC_DISPATCH_SELECTOR)) && context.environment.calldata.length > 0) {
+    functionSelector = FunctionSelector.fromField(context.environment.calldata[0]);
+  }
```

Comment:
I know we're leaking app-level information to the AVM, but considering this is really for "debuggability", this might be okay. It's concept leakage, but only for the purpose of useful error messages.